Add Spawner.progress_ready_hook for customizing the ready event

kreuzert opened the pull request:
discussion: https://github.com/jupyterhub/jupyterhub/issues/4447

minrk commented:
Looks sensible to me, thank you! Can you add a test for this in `test_api.py` around the other `test_progress_` tests?

kreuzert commented:
I've added three tests (async, sync, exception) for `progress_ready_hook` and switched to `deepcopy` to create the `original_ready_event`. Otherwise, modifications of `ready_event` in `progress_ready_hook` would not have been reverted.
  
Please let me know what you think about it.

minrk commented on jupyterhub/apihandlers/users.py:
```diff
@@ -710,19 +712,32 @@ async def get(self, user_name, server_name=''):
+                'html_message': 'Server ready at <a href="{0}">{0}</a>'.format(url),
+                'url': url,
+            }
+            original_ready_event = deepcopy(ready_event)
```
deepcopy is fine here, but I don't think it is strictly required _in this case_ because all of the values are immutable scalars. So there isn't really a difference between a shallow and deep copy of this dict:

```python
In [8]: d = {"html_message": "Server ready"}

In [9]: d2 = d.copy()

In [10]: d2["html_message"] += " extra"

In [11]: d2["extra_key"] = "ok"

In [12]: d
Out[12]: {'html_message': 'Server ready'}

In [13]: d2
Out[13]: {'html_message': 'Server ready extra', 'extra_key': 'ok'}
```

deepcopy only makes a difference when the _values_ of the dict are themselves mutable, like lists, dicts, etc.

minrk commented on jupyterhub/spawner.py:
```diff
@@ -840,6 +840,28 @@ def _deprecate_percent_u(self, proposal):
+
+        Example::
+
+            from subprocess import check_call
```
```suggestion
```

minrk commented on jupyterhub/spawner.py:
```diff
@@ -840,6 +840,28 @@ def _deprecate_percent_u(self, proposal):
+    progress_ready_hook = Any(
+        help="""
+        An optional hook function that you can implement to modify the
+        ready event, which will be shown to the user once a service
```
```suggestion
        ready event, which will be shown to the user on the spawn progress page when their server
```

minrk reviewed (commented):
Looks great, thanks for the test! It's still marked as draft, is there anything you'd like to add?

My only suggestions are in the hook docstring.

kreuzert commented:
Thanks for your feedback, I've updated it. From my side, we can merge it.

minrk commented:
Thanks!
